signatures.md

```diff
@@ -52,163 +49,350 @@ Note: `Message` is used here as the object being signed, but this interface shou
+Where h(msg) is the output of a cryptographic hash function and sigma is the notation for BLS signature.
+
+## Aggregating Signed Messages
+
+Pairing-based signatures allow a user to represent a collection of signatures as a single element in the base group. In other words, we can take a collection of signatures and "add" them together into a single value that is the same size as a single signature. We call this process Signature Aggregation. The result of this operation is called the Aggregate Signature which has the same general properties as any other pairing-based Signature. Some security [considerations](https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html).
+
+### Aggregation Process (taken from [IETF BLS Signature Scheme - Aggregate](https://tools.ietf.org/html/draft-boneh-bls-signature-00#section-2.5) )
 
-`r` = Compressed elliptic curve point (x-coordinate) of size 32 bytes
+```go
+Input: (PK_1, sigma_1), ..., (PK_n, sigma_n)    
+Output: sigma
+
+   1.  Output sigma = sigma_1 + sigma_2 + ... + sigma_n
+```
 
-`recovery` = Information needed to recover a public key from `sig`. 
+Where `sigma_i` is a Signature with associated public key  `PK_i`. `sigma` is the Aggregate Signature and will be represented by `SignatureBytes`. `+` is point addition as defined by the cryptographic group.
 
-- LSB(0) = parity of y-coordinate of r
-- LSB(1) = overflow indicator
+## Verifying Signed Messages
 
-`indicator` = a 2 byte formatting indicator
+In order to communicate to `Verify` what kind of verification algorithm to use, Verify must accept `SignedMessage` as input. 
 
-From: <https://github.com/bitcoin-core/secp256k1/blob/314a61d72474aa29ff4afba8472553ad91d88e9d/src/ecdsa_impl.h#L177>
+**Determining Signature Type**
 
-**Signed Message**
+`SignedMessage.Address` will inform `Verify` if the `SignedMessage.SignatureBytes` is a BLS or ECDSA signature. `Verify` will infer this data from the `Address.protocol` field. 
 
+```go
+Input: Address
+Output: "BLS", "ECDSA", or "INVALID"
+   
+   1.  If Address.protocol = 0 output "BLS" and stop
+   2.  If Address.protocol = 1 output "ECDSA" and stop
+   3.  If Address.protocol = 3 output "BLS" and stop
+	 4.  If Address.protocol != {0 OR 1 OR 3} output "INVALID"
 ```
-Type SignedMessage Struct {
 
-    Message bytes `JSON`   
+**Determining if the signature is a Signature Aggregate**
 
-    Signature bytes `JSON` 
-    
-}
+If there is more then one message in `SignedMessage.Msgs` then the `SignedMessage` is assumed to be a signature aggregate. If `Address` is empty then `Verify` will assume the `SignedMessage` is a signature aggregate. The procedure below uses `len()` to indicate a function that returns the number of elements in an array.
+
+```go
+Input: SignedMessage
+Output: "Aggregate" or "INVALID"
+	    
+	 1.  If SignedMessage.Address[i].protocol != {0 OR 3} for all i output "INVALID" and stop
+	 2.  If len(SignedMessage.Msgs) > 1 output "Aggregate" else output "INVALID"
```

Comment:
Would the case `len(SignedMessage.Msgs) == 1` indicate the the `SignedMessage` contains a ECDSA signature?

Reply:
What about the case of no addresses and one message? Is an aggregate of one explicitly invalid? While that's what the pseudocode implied, explicit clarification would help.

What about the case of multiple addresses and multiple messages? I'm guessing this means an aggregate (but should it check the lengths are equal)? An enumeration of all possible cases of {1, many}-messages ⨯ {0, many}-addresses would help (perhaps in the prose above).

Reply:
@frrist: yes. ECDSA signatures can't be aggregated in this scheme so if we encounter them we should error. Maybe "INVALID" is not the right wording. We should probably specify input validation. Alex alludes to this below. 

@anorth: Good point. Request someone from go-filecoin to work on this (input validation).

Reply:
Added note about ECDSA not being aggregatable in the "Aggregating Signed Messages" section

Reply:
I would love us to work on input validation, but we need you to specify what inputs should be valid. Please enumerate all combinations that are semantically valid and explicitly state that other combinations are meaningless.